contracts/Curation.sol

```diff
@@ -0,0 +1,415 @@
+pragma solidity 0.4.18;
+
+import "@aragon/os/contracts/apps/AragonApp.sol";
+import "@aragon/os/contracts/lib/misc/Migrations.sol";
+import "@aragon/os/contracts/lib/zeppelin/math/SafeMath.sol";
+import "@aragon/os/contracts/lib/zeppelin/math/SafeMath64.sol";
+
+import "@aragon/apps-staking/contracts/interfaces/IStaking.sol";
+import "./interfaces/IRegistry.sol";
+import "./interfaces/IVoting.sol";
+
+
+contract Curation is AragonApp {
+    using SafeMath for uint256;
+    using SafeMath64 for uint64;
+
+    uint64 constant public MAX_UINT64 = uint64(-1);
+    uint256 constant public PCT_BASE = 10 ** 18; // 0% = 0; 1% = 10^16; 100% = 10^18
+    bytes32 constant public MIN_DEPOSIT_PARAM = keccak256("MIN_DEPOSIT_PARAM");
+    bytes32 constant public APPLY_STAGE_LEN_PARAM = keccak256("APPLY_STAGE_LEN_PARAM");
+    bytes32 constant public DISPENSATION_PCT_PARAM = keccak256("DISPENSATION_PCT_PARAM");
+
+    IRegistry public registry;
+    IStaking public staking;
+    IVoting public voting;
+    uint256 public minDeposit;
+    uint64 public applyStageLen;
+    uint256 public dispensationPct;
+
+    bytes32 constant public CHANGE_PARAMS_ROLE = keccak256("CHANGE_PARAMS_ROLE");
+    bytes32 constant public CHANGE_VOTING_APP_ROLE = keccak256("CHANGE_VOTING_APP_ROLE");
+
+    // TODO!!!
+    enum TimeUnit { Blocks, Seconds }
+
+    struct Application {
+        address applicant;
+        uint64 date;
+        bool registered;
+        bytes data;
+        uint256 amount;
+        uint256 lockId;
+    }
+
+    struct Challenge {
+        address challenger;
+        uint64 date;
+        bool resolved;
+        uint256 amount;
+        uint256 lockId;
```

Comment:
We may run into issues with `lockIds` changing as when you remove a lock in the staking app the ids can change: https://github.com/aragon/aragon-apps/blob/b968e3ec9ab70cf6cfbfa5ebca8a58460669e7d1/future-apps/staking/contracts/Staking.sol#L173-L178

If a user has an unrelated lock with id = 0 and then creates a lock for creating a challenge with id = 1, if the lock 0 gets unlocked before the challenge is resolved, lock 1 will be lock 0 and unlocking will fail as there is no lock 1.

I think this issue may apply to PLCR voting as well. We may need to rethink how locks get stored in the Staking app.

Reply:
Yes, since lock ids are exposed to the outside world, they should be immutable.

Reply:
Created https://github.com/aragon/aragon-apps/issues/406 for this